From 5f181dc1fc1710d1b9426ce545a342966989640a Mon Sep 17 00:00:00 2001
From: Albin B <albin.bala.krishnan@intel.com>
Date: Tue, 15 Sep 2015 15:27:53 +0530
Subject: [PATCH 23/28] [EM/bq24192]: Add oneshot to serialize the interrupt
 processing

To serialize the charger irq's via the soc gpio use
IRQF_ONESHOT when register for irq. Also use devm for irq
registration to get the irq freed automatically.

Change-Id: Ic3bd4d7ad63e184cf48f381a44367b6a90684639
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-9421
Signed-off-by: Albin B <albin.bala.krishnan@intel.com>
Reviewed-on: https://android.intel.com:443/429501
(cherry picked from commit 238168e443093e752e41c9c286f4c7163de4f8b1)
Reviewed-on: https://android.intel.com:443/438661
---
 drivers/power/bq24192_charger.c | 8 +++-----
 1 file changed, 3 insertions(+), 5 deletions(-)

diff --git a/drivers/power/bq24192_charger.c b/drivers/power/bq24192_charger.c
index 308b5a1..458b4a8 100644
--- a/drivers/power/bq24192_charger.c
+++ b/drivers/power/bq24192_charger.c
@@ -2392,9 +2392,10 @@ static int bq24192_probe(struct i2c_client *client,
 		dev_err(&chip->client->dev,
 			"chgr_int_n GPIO is not available\n");
 	} else {
-		ret = request_threaded_irq(chip->irq,
+		ret = devm_request_threaded_irq(&bq24192_client->dev, chip->irq,
 				bq24192_irq_isr, bq24192_irq_thread,
-				IRQF_TRIGGER_FALLING, "BQ24192", chip);
+				IRQF_ONESHOT | IRQF_TRIGGER_FALLING,
+				"BQ24192", chip);
 		if (ret) {
 			dev_err(&bq24192_client->dev,
 				"failed to register irq for pin %d\n",
@@ -2531,9 +2532,6 @@ static int bq24192_remove(struct i2c_client *client)
 	if (!chip->pdata->slave_mode)
 		power_supply_unregister(&chip->usb);
 
-	if (chip->irq > 0)
-		free_irq(chip->irq, chip);
-
 	i2c_set_clientdata(client, NULL);
 	wake_lock_destroy(&chip->wakelock);
 	kfree(chip);
-- 
1.9.1

